File: rb/test/api_entities_cluster_group_entity_test.rb
# ApiEntitiesClusterGroup entity test

require "minitest/autorun"
require "json"
require_relative "../Gitlab_sdk"
require_relative "runner"

class ApiEntitiesClusterGroupEntityTest < Minitest::Test
  def test_create_instance
    testsdk = GitlabSDK.test(nil, nil)
    ent = testsdk.ApiEntitiesClusterGroup(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = api_entities_cluster_group_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create", "update", "load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "api_entities_cluster_group." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set GITLAB_TEST_API_ENTITIES_CLUSTER_GROUP_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    api_entities_cluster_group_ref01_ent = client.ApiEntitiesClusterGroup(nil)
    api_entities_cluster_group_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.api_entities_cluster_group"), "api_entities_cluster_group_ref01"))
    api_entities_cluster_group_ref01_data["group_id"] = setup[:idmap]["group01"]

    api_entities_cluster_group_ref01_data_result = api_entities_cluster_group_ref01_ent.create(api_entities_cluster_group_ref01_data, nil)
    api_entities_cluster_group_ref01_data = Helpers.to_map(api_entities_cluster_group_ref01_data_result)
    assert !api_entities_cluster_group_ref01_data.nil?
    assert !api_entities_cluster_group_ref01_data["id"].nil?

    # UPDATE
    api_entities_cluster_group_ref01_data_up0_up = {
      "id" => api_entities_cluster_group_ref01_data["id"],
      "group_id" => setup[:idmap]["group_id"],
    }

    api_entities_cluster_group_ref01_markdef_up0_name = "cluster_type"
    api_entities_cluster_group_ref01_markdef_up0_value = "Mark01-api_entities_cluster_group_ref01_#{setup[:now]}"
    api_entities_cluster_group_ref01_data_up0_up[api_entities_cluster_group_ref01_markdef_up0_name] = api_entities_cluster_group_ref01_markdef_up0_value

    api_entities_cluster_group_ref01_resdata_up0_result = api_entities_cluster_group_ref01_ent.update(api_entities_cluster_group_ref01_data_up0_up, nil)
    api_entities_cluster_group_ref01_resdata_up0 = Helpers.to_map(api_entities_cluster_group_ref01_resdata_up0_result)
    assert !api_entities_cluster_group_ref01_resdata_up0.nil?
    assert_equal api_entities_cluster_group_ref01_resdata_up0["id"], api_entities_cluster_group_ref01_data_up0_up["id"]
    assert_equal api_entities_cluster_group_ref01_resdata_up0[api_entities_cluster_group_ref01_markdef_up0_name], api_entities_cluster_group_ref01_markdef_up0_value

    # LOAD
    api_entities_cluster_group_ref01_match_dt0 = {
      "id" => api_entities_cluster_group_ref01_data["id"],
    }
    api_entities_cluster_group_ref01_data_dt0_loaded = api_entities_cluster_group_ref01_ent.load(api_entities_cluster_group_ref01_match_dt0, nil)
    api_entities_cluster_group_ref01_data_dt0_load_result = Helpers.to_map(api_entities_cluster_group_ref01_data_dt0_loaded)
    assert !api_entities_cluster_group_ref01_data_dt0_load_result.nil?
    assert_equal api_entities_cluster_group_ref01_data_dt0_load_result["id"], api_entities_cluster_group_ref01_data["id"]

  end
end

def api_entities_cluster_group_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "api_entities_cluster_group", "ApiEntitiesClusterGroupTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = GitlabSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["api_entities_cluster_group01", "api_entities_cluster_group02", "api_entities_cluster_group03", "group01", "group02", "group03", "cluster01", "cluster02", "cluster03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["GITLAB_TEST_API_ENTITIES_CLUSTER_GROUP_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "GITLAB_TEST_API_ENTITIES_CLUSTER_GROUP_ENTID" => idmap,
    "GITLAB_TEST_LIVE" => "FALSE",
    "GITLAB_TEST_EXPLAIN" => "FALSE",
    "GITLAB_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["GITLAB_TEST_API_ENTITIES_CLUSTER_GROUP_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end
  if idmap_resolved["group_id"].nil?
    idmap_resolved["group_id"] = idmap_resolved["group01"]
  end

  if env["GITLAB_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["GITLAB_APIKEY"],
      },
      extra || {},
    ])
    client = GitlabSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["GITLAB_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["GITLAB_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: go/test/api_entities_cluster_group_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/gitlab-sdk/go"
	"github.com/voxgig-sdk/gitlab-sdk/go/core"

	vs "github.com/voxgig-sdk/gitlab-sdk/go/utility/struct"
)

func TestApiEntitiesClusterGroupEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.ApiEntitiesClusterGroup(nil)
		if ent == nil {
			t.Fatal("expected non-nil ApiEntitiesClusterGroupEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := api_entities_cluster_groupBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "update", "load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "api_entities_cluster_group." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GITLAB_TEST_API_ENTITIES_CLUSTER_GROUP_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		apiEntitiesClusterGroupRef01Ent := client.ApiEntitiesClusterGroup(nil)
		apiEntitiesClusterGroupRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "api_entities_cluster_group"}), "api_entities_cluster_group_ref01"))
		apiEntitiesClusterGroupRef01Data["group_id"] = setup.idmap["group01"]

		apiEntitiesClusterGroupRef01DataResult, err := apiEntitiesClusterGroupRef01Ent.Create(apiEntitiesClusterGroupRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		apiEntitiesClusterGroupRef01Data = core.ToMapAny(entityData(apiEntitiesClusterGroupRef01DataResult))
		if apiEntitiesClusterGroupRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if apiEntitiesClusterGroupRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// UPDATE
		apiEntitiesClusterGroupRef01DataUp0Up := map[string]any{
			"id": apiEntitiesClusterGroupRef01Data["id"],
			"group_id": setup.idmap["group_id"],
		}

		apiEntitiesClusterGroupRef01MarkdefUp0Name := "cluster_type"
		apiEntitiesClusterGroupRef01MarkdefUp0Value := fmt.Sprintf("Mark01-api_entities_cluster_group_ref01_%d", setup.now)
		apiEntitiesClusterGroupRef01DataUp0Up[apiEntitiesClusterGroupRef01MarkdefUp0Name] = apiEntitiesClusterGroupRef01MarkdefUp0Value

		apiEntitiesClusterGroupRef01ResdataUp0Result, err := apiEntitiesClusterGroupRef01Ent.Update(apiEntitiesClusterGroupRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		apiEntitiesClusterGroupRef01ResdataUp0 := core.ToMapAny(entityData(apiEntitiesClusterGroupRef01ResdataUp0Result))
		if apiEntitiesClusterGroupRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if apiEntitiesClusterGroupRef01ResdataUp0["id"] != apiEntitiesClusterGroupRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if apiEntitiesClusterGroupRef01ResdataUp0[apiEntitiesClusterGroupRef01MarkdefUp0Name] != apiEntitiesClusterGroupRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", apiEntitiesClusterGroupRef01MarkdefUp0Name, apiEntitiesClusterGroupRef01ResdataUp0[apiEntitiesClusterGroupRef01MarkdefUp0Name])
		}

		// LOAD
		apiEntitiesClusterGroupRef01MatchDt0 := map[string]any{
			"id": apiEntitiesClusterGroupRef01Data["id"],
		}
		apiEntitiesClusterGroupRef01DataDt0Loaded, err := apiEntitiesClusterGroupRef01Ent.Load(apiEntitiesClusterGroupRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		apiEntitiesClusterGroupRef01DataDt0LoadResult := core.ToMapAny(entityData(apiEntitiesClusterGroupRef01DataDt0Loaded))
		if apiEntitiesClusterGroupRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if apiEntitiesClusterGroupRef01DataDt0LoadResult["id"] != apiEntitiesClusterGroupRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func api_entities_cluster_groupBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "api_entities_cluster_group", "ApiEntitiesClusterGroupTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read api_entities_cluster_group test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse api_entities_cluster_group test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"api_entities_cluster_group01", "api_entities_cluster_group02", "api_entities_cluster_group03", "group01", "group02", "group03", "cluster01", "cluster02", "cluster03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GITLAB_TEST_API_ENTITIES_CLUSTER_GROUP_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GITLAB_TEST_API_ENTITIES_CLUSTER_GROUP_ENTID": idmap,
		"GITLAB_TEST_LIVE":      "FALSE",
		"GITLAB_TEST_EXPLAIN":   "FALSE",
		"GITLAB_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["GITLAB_TEST_API_ENTITIES_CLUSTER_GROUP_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}
	// Add group_id alias for update test.
	if idmapResolved["group_id"] == nil {
		idmapResolved["group_id"] = idmapResolved["group01"]
	}

	if env["GITLAB_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["GITLAB_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewGitlabSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GITLAB_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GITLAB_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
